### PR TITLE
8304502: Classfile API class hierarchy makes assumptions when class is not resolved

### PR DESCRIPTION
**Pull request by asotona**
Classfile API class hierarchy makes assumptions when class is not resolved and that may lead to silent generation of invalid stack maps. Only debug-level log information of case is actually provided.

Proposed patch throws IllegalArgumentException when the class is not resolved instead.

Thanks for review.

Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304502](https://bugs.openjdk.org/browse/JDK-8304502): Classfile API class hierarchy makes assumptions when class is not resolved

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13099/head:pull/13099` \
`$ git checkout pull/13099`

Update a local copy of the PR: \
`$ git checkout pull/13099` \
`$ git pull https://git.openjdk.org/jdk.git pull/13099/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13099`

View PR using the GUI difftool: \
`$ git pr show -t 13099`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13099.diff">https://git.openjdk.org/jdk/pull/13099.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java [hunk 1]
```diff
@@ -176,7 +168,8 @@ public static final class StaticClassHierarchyResolver implements ClassHierarchy
         private final Map<ClassDesc, ClassHierarchyInfo> map;
 
         public StaticClassHierarchyResolver(Collection<ClassDesc> interfaceNames, Map<ClassDesc, ClassDesc> classToSuperClass) {
-            map = new HashMap<>(interfaceNames.size() + classToSuperClass.size());
+            map = new HashMap<>(interfaceNames.size() + classToSuperClass.size() + 1);
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java:171]: Should use HashMap.newHashMap instead
```suggestion
            map = HashMap.newHashMap(interfaceNames.size() + classToSuperClass.size() + 1);
```

**Review by jaikiran — APPROVED** [review]
Looks OK to me. Just a minor comment, which I've added inline.

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java [hunk 2]
```diff
@@ -176,7 +168,8 @@ public static final class StaticClassHierarchyResolver implements ClassHierarchy
         private final Map<ClassDesc, ClassHierarchyInfo> map;
 
         public StaticClassHierarchyResolver(Collection<ClassDesc> interfaceNames, Map<ClassDesc, ClassDesc> classToSuperClass) {
-            map = new HashMap<>(interfaceNames.size() + classToSuperClass.size());
+            map = HashMap.newHashMap(interfaceNames.size() + classToSuperClass.size() + 1);
+            map.put(ConstantDescs.CD_Object, new ClassHierarchyInfo(ConstantDescs.CD_Object, false, null));
```
[the remark this answers is not in the document]
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java]: fixed, thanks.

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**